Fix to `F.max_pooling_2d` test input would make the test fail

https://github.com/chainer/chainer/blob/v7.0.0b2/tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py#L52
```
        numpy.random.shuffle(x)
```

This line should be `numpy.random.shuffle(x.ravel())` because `shuffle` only shuffles along the first axis.
But that fix would cause the test to fail (`test_double_backward` with cuDNN enabled).

`pytest -m 'cudnn' -v -rfEX --tb=short tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py -k 'test_double_backward'`

```
E   chainer.testing.function_link.FunctionTestError: Parameterized test failed.
E
E   Base test method: TestMaxPooling2D_use_chainerx_false__chainerx_device_None__use_cuda_true__cuda_device_0__use_cudnn_always__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never.test_doubl$_backward
E   Test parameters:
E     contiguous: C
E     cover_all: False
E     dtype: <class 'numpy.float64'>
E
E
E   (caused by)
E   FunctionTestError: double backward is not implemented correctly

:
:

E   gradients (numeric):  -0.005565301559232316
E   gradients (backward): 0.14605092276952303
E
E   x: numeric gradient, y: backward gradient
E   Not equal to tolerance rtol=0.001, atol=0.0001
E
E   Mismatch: 100%
E   Max absolute difference: 0.15161622
E   Max relative difference: 1.03810521
E    x: array(-0.005565)
E    y: array(0.146051)
E
E   assert_allclose failed:
E     shape: () ()
E     dtype: float64 float64
E     i: (0,)
E     x[i]: -0.005565301559232316
E     y[i]: 0.14605092276952303
E     relative error[i]: 1.0381052132619162
E     absolute error[i]: 0.15161622432875535
E     relative tolerance * |y[i]|: 0.00014605092276952304
E     absolute tolerance: 0.0001
E     total tolerance: 0.00024605092276952306
E   x: -0.0055653
E   y: 0.14605092
```

--- a/tests/chainer_tests/functions_tests/pooling_tests/pooling_nd_helper.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/pooling_nd_helper.py
@@ -1,7 +1,10 @@
 import itertools
+
+import numpy
 import six
 
 from chainer import testing
+import chainer.utils
 
 
 def pooling_patches(dims, ksize, stride, pad, cover_all):
@@ -19,6 +22,13 @@ def pooling_patches(dims, ksize, stride, pad, cover_all):
     return [tuple(slice(max(x, 0), min(x + k, d))
                   for (x, d, k) in six.moves.zip(xs, dims, ksize))
             for xs in xss]
+
+
+def shuffled_linspace(shape, dtype):
+    size = chainer.utils.size_of_shape(shape)
+    x = numpy.random.permutation(size) + numpy.random.uniform(0.3, 0.7, size)
+    x = (2 * x / max(1, size) - 1).astype(dtype)
+    return x.reshape(shape)
 
 
 testing.run_module(__name__, __file__)

--- a/tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py
@@ -9,6 +9,7 @@ from chainer import functions
 from chainer import testing
 from chainer.testing import attr
 from chainer.testing import backend
+from chainer_tests.functions_tests.pooling_tests import pooling_nd_helper
 
 
 _inject_backend_tests = backend.inject_backend_tests(
@@ -57,10 +58,7 @@ class TestMaxPooling2D(testing.FunctionTestCase):
                 'atol': 1e-4, 'rtol': 1e-3}
 
     def generate_inputs(self):
-        x = numpy.arange(2 * 3 * 4 * 3, dtype=self.dtype).reshape(2, 3, 4, 3)
-        numpy.random.shuffle(x)
-        x = 2 * x / x.size - 1
-        return x,
+        return pooling_nd_helper.shuffled_linspace((2, 3, 4, 3), self.dtype),
 
     def forward_expected(self, inputs):
         x, = inputs
@@ -136,9 +134,8 @@ class TestMaxPooling2DCudnnCall(unittest.TestCase):
 
 class TestMaxPooling2DIndices(unittest.TestCase):
     def setUp(self):
-        self.x = numpy.arange(
-            2 * 3 * 4 * 4, dtype=numpy.float32).reshape(2, 3, 4, 4)
-        numpy.random.shuffle(self.x)
+        self.x = pooling_nd_helper.shuffled_linspace(
+            (2, 3, 4, 4), numpy.float32)
 
     def _check(self, x):
         out, indices = functions.max_pooling_2d(

--- a/tests/chainer_tests/functions_tests/pooling_tests/test_roi_average_align_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_roi_average_align_2d.py
@@ -9,6 +9,7 @@ from chainer import gradient_check
 from chainer import testing
 from chainer.testing import attr
 from chainer.testing import condition
+from chainer_tests.functions_tests.pooling_tests import pooling_nd_helper
 
 
 def _pair(x):
@@ -35,11 +36,8 @@ class TestROIAlign2D(unittest.TestCase):
     def setUp(self):
         N = 3
         n_channels = 3
-        self.x = numpy.arange(
-            N * n_channels * 12 * 8,
-            dtype=numpy.float32).reshape((N, n_channels, 12, 8))
-        numpy.random.shuffle(self.x)
-        self.x = 2 * self.x / self.x.size - 1
+        self.x = pooling_nd_helper.shuffled_linspace(
+            (N, n_channels, 12, 8), numpy.float32)
         self.rois = numpy.array([
             [1, 1, 6, 6],
             [2, 6, 11, 7],

--- a/tests/chainer_tests/functions_tests/pooling_tests/test_roi_average_pooling_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_roi_average_pooling_2d.py
@@ -9,6 +9,7 @@ from chainer import gradient_check
 from chainer import testing
 from chainer.testing import attr
 from chainer.utils import collections_abc
+from chainer_tests.functions_tests.pooling_tests import pooling_nd_helper
 
 
 def _pair(x):
@@ -29,11 +30,8 @@ class TestROIAveragePooling2D(unittest.TestCase):
     def setUp(self):
         N = 3
         n_channels = 3
-        self.x = numpy.arange(
-            N * n_channels * 12 * 8,
-            dtype=numpy.float32).reshape((N, n_channels, 12, 8))
-        numpy.random.shuffle(self.x)
-        self.x = (2 * self.x / self.x.size - 1).astype(self.dtype)
+        self.x = pooling_nd_helper.shuffled_linspace(
+            (N, n_channels, 12, 8), self.dtype)
         self.rois = numpy.array([
             [1, 1, 7, 7],
             [2, 6, 12, 8],

--- a/tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_align_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_align_2d.py
@@ -9,6 +9,7 @@ from chainer import gradient_check
 from chainer import testing
 from chainer.testing import attr
 from chainer.testing import condition
+from chainer_tests.functions_tests.pooling_tests import pooling_nd_helper
 
 
 def _pair(x):
@@ -35,11 +36,8 @@ class TestROIMaxAlign2D(unittest.TestCase):
     def setUp(self):
         N = 3
         n_channels = 3
-        self.x = numpy.arange(
-            N * n_channels * 12 * 8,
-            dtype=numpy.float32).reshape((N, n_channels, 12, 8))
-        numpy.random.shuffle(self.x)
-        self.x = 2 * self.x / self.x.size - 1
+        self.x = pooling_nd_helper.shuffled_linspace(
+            (N, n_channels, 12, 8), numpy.float32)
         self.rois = numpy.array([
             [1, 1, 6, 6],
             [6, 2, 7, 11],

--- a/tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_pooling_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_pooling_2d.py
@@ -8,6 +8,7 @@ from chainer import functions
 from chainer import gradient_check
 from chainer import testing
 from chainer.testing import attr
+from chainer_tests.functions_tests.pooling_tests import pooling_nd_helper
 
 
 def _pair(x):
@@ -29,11 +30,8 @@ class TestROIMaxPooling2D(unittest.TestCase):
     def setUp(self):
         N = 3
         n_channels = 3
-        self.x = numpy.arange(
-            N * n_channels * 12 * 8,
-            dtype=numpy.float32).reshape((N, n_channels, 12, 8))
-        numpy.random.shuffle(self.x)
-        self.x = (2 * self.x / self.x.size - 1).astype(self.dtype)
+        self.x = pooling_nd_helper.shuffled_linspace(
+            (N, n_channels, 12, 8), self.dtype)
         self.rois = numpy.array([
             [1, 1, 7, 7],
             [2, 6, 12, 8],

--- a/tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py
@@ -8,6 +8,7 @@ from chainer import functions
 from chainer import gradient_check
 from chainer import testing
 from chainer.testing import attr
+from chainer_tests.functions_tests.pooling_tests import pooling_nd_helper
 
 
 @testing.parameterize(*testing.product({
@@ -18,11 +19,8 @@ class TestROIPooling2D(unittest.TestCase):
     def setUp(self):
         N = 3
         n_channels = 3
-        self.x = numpy.arange(
-            N * n_channels * 12 * 8,
-            dtype=numpy.float32).reshape((N, n_channels, 12, 8))
-        numpy.random.shuffle(self.x)
-        self.x = (2 * self.x / self.x.size - 1).astype(self.dtype)
+        self.x = pooling_nd_helper.shuffled_linspace(
+            (N, n_channels, 12, 8), self.dtype)
         self.rois = numpy.array([
             [0, 1, 1, 6, 6],
             [2, 6, 2, 7, 11],

--- a/tests/chainer_tests/functions_tests/pooling_tests/test_spatial_pyramid_pooling_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_spatial_pyramid_pooling_2d.py
@@ -10,6 +10,7 @@ from chainer import testing
 from chainer.testing import attr
 from chainer.testing import condition
 from chainer.utils import type_check
+from chainer_tests.functions_tests.pooling_tests import pooling_nd_helper
 
 
 @testing.parameterize(*testing.product_dict(
@@ -32,12 +33,7 @@ class TestSpatialPyramidPooling2D(unittest.TestCase):
         # To avoid instability of numerical gradient, use enough different
         # values.
         shape = (self.n, self.c, self.h, self.w)
-        size = numpy.prod(shape)
-        self.x = numpy.arange(size, dtype=self.dtype).reshape(shape)
-        numpy.random.shuffle(self.x)
-        self.x += numpy.random.uniform(
-            0.4, 0.6, shape).astype(self.dtype)
-        self.x /= size
+        self.x = pooling_nd_helper.shuffled_linspace(shape, self.dtype)
 
         self.one = numpy.ones(
             (self.n, self.c, self.h, self.w)).astype(self.dtype)

--- a/tests/chainer_tests/functions_tests/pooling_tests/test_unpooling_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_unpooling_2d.py
@@ -9,6 +9,7 @@ from chainer import functions
 from chainer import gradient_check
 from chainer import testing
 from chainer.testing import attr
+from chainer_tests.functions_tests.pooling_tests import pooling_nd_helper
 
 
 @testing.parameterize(*testing.product_dict(
@@ -36,11 +37,8 @@ class TestUnpooling2D(unittest.TestCase):
         self.N = 2
         self.n_channels = 3
         inh, inw = 2, 1
-        self.x = numpy.arange(
-            self.N * self.n_channels * inh * inw,
-            dtype=self.dtype).reshape(self.N, self.n_channels, inh, inw)
-        numpy.random.shuffle(self.x)
-        self.x = 2 * self.x / self.x.size - 1
+        self.x = pooling_nd_helper.shuffled_linspace(
+            (self.N, self.n_channels, inh, inw), self.dtype)
 
         self.ksize = 2
         outh, outw = self.outsize or self.expected_outsize
@@ -159,11 +157,8 @@ class TestIntegerScaleUnpooling2D(unittest.TestCase):
         self.N = 2
         self.n_channels = 3
         inh, inw = self.insize
-        self.x = numpy.arange(
-            self.N * self.n_channels * inh * inw,
-            dtype=self.dtype).reshape(self.N, self.n_channels, inh, inw)
-        numpy.random.shuffle(self.x)
-        self.x = 2 * self.x / self.x.size - 1
+        self.x = pooling_nd_helper.shuffled_linspace(
+            (self.N, self.n_channels, inh, inw), self.dtype)
 
         outh, outw = self.outsize or self.expected_outsize
         self.gy = numpy.random.uniform(

